src/controlsocket.py

```diff
@@ -0,0 +1,227 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# Licensed under the GPLv3, see LICENSE file for details.
+
+import email.message
+import email.parser
+import http.client
+import json
+import logging
+import socket
+import sys
+import urllib.error
+import urllib.parse
+import urllib.request
+from typing import (
+    Any,
+    Dict,
+    Generator,
+    Literal,
+    Optional,
+    Union,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class ControlSocketClient:
```

Comment:
I forget what we discussed previously, but seeing the charm is using `ops` anyway, if you wanted you could take a shortcut and implement this using `ops.pebble.Client` (via composition, not inheritance). You'd depend on the private method `Client._request`, but that's unlikely to change anytime soon, and if you're pinning the ops version you'd be safe anyway. It would save reimplementing this stuff (and copying and maintaining the tests here too).

That way you'd only have to implement the two new metrics methods.

Reply:
hmm, I kinda prefer copy pasting it because it gives us the freedom to change this independently of what happens in Pebble. Considering this is a very new feature, we don't know how it will evolve.